API: Restructure WSIReader to allow support for additional formats

shaneahmed opened the pull request:
- Restructure WSIReader as parent class to allow support to read whole slide images in other formats.
- Add OpenSlideWSIReader
- Add tests for OpenSlideWSIReader
- Add docs for OpenSlideWSIReader

simongraham commented on tiatoolbox/dataloader/slide_info.py:
```diff
@@ -1,20 +1,21 @@
     Args:
         input_path (str): Path to whole slide image
         output_dir (str): Path to output directory to save the output
+        verbose(bool): Print output, default=True
```
space after verbose

shaneahmed replied:
Fixed in this commit https://github.com/TIA-Lab/tiatoolbox/commit/3ac21d1b573dc9f272a8527ef14e866b52c9c631

simongraham commented on tiatoolbox/dataloader/slide_info.py:
```diff
@@ -34,16 +35,18 @@ def slide_info(input_path, output_dir=None):
     else:
-        print("File type not supported")
-        info = None
+        raise FileNotSupported
```
Doing it this way will raise a generic error - do we not want to also give the filename for which it failed on?

shaneahmed replied:
Updated it in this commit https://github.com/TIA-Lab/tiatoolbox/commit/3ac21d1b573dc9f272a8527ef14e866b52c9c631

simongraham commented on tiatoolbox/utils/exceptions.py:
```diff
@@ -0,0 +1,8 @@
+"""Custom Errors and Exceptions for TIAToolbox"""
+
+
+class FileNotSupported(Exception):
```
instead - word as:  'File format is not supported'

shaneahmed replied:
Updated in this commit https://github.com/TIA-Lab/tiatoolbox/commit/dd194c4ee46e2bc4f36c73138f8f03515037b1fd

shaneahmed commented:
> Overall the PR is good and I like how we subclass the WSIReader. The next step can be to subclass with a JP2Reader. Why do we use double underscore for `__slide_info__`? This is normally reserved for inbuilt python magic methods - we can instead just use double leading underscore to denote a private method. For the exceptions, do you think we should instead simply use a function, with the filename as an argument. This way we can give a more specific error message. Just something to consider.
> 
> Other than that, a couple of very minor things to fix before merging. I will approve, but please look into making those small changes before making the merge. Also, for the module string added to the top of a script, should we not add a space in between the module string and first library import?

Thanks Simon. I have addressed your concerns on ___slide_info__ and error message. Regarding space after docstring at the top of module, let's discuss this on the channel and then make it consistent after adding to OneNote.

John-P commented on tiatoolbox/cli.py:
```diff
@@ -119,22 +126,33 @@ def slide_info(wsi_input, output_dir, file_types, mode, workers=None):
 def read_region(wsi_input, region, level, output_path, mode):
     """Reads a region in an whole slide image as specified"""
-    if all(region):
+    if not region:
```
Is this checking if that region is not None? I would consider changing it to `if region is not None:` if so as it is more clear.

shaneahmed replied:
Its to check empty tuple "()".
